In [88]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_val_score

# Part 1: Load the dataset

In [78]:
# Load the dataset (load remotely or locally)
df = pd.read_csv("iris.csv")

# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
print(df.head(15))
print(f"\n Shape of the dataset: {df.shape}")
print(f"\n {df.describe()}")
print(f"\n Columns: {df.columns}")


    sepal_length  sepal_width  petal_length  petal_width species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14           5.8         

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

The data contains the measurements of the different flower parts for three flower species: setosa, versicolor, and virginica. The label is the flower species, and the features are the sepal length, sepal width, petal length, and petal width. The mapping of the labels to the actual classes is 0 = setosa, 1 = versicolor, 2 = virginica.


# Part 2: Split the dataset into train and test

In [79]:
# Take the dataset and split it into our features (X) and label (y)
X = df.drop('species', axis=1)
y = df['species']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)

#changed to 80% train, 20% test as per professor's suggestion for varying train test split 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=10)

# Part 3: Logistic Regression

In [80]:
# i. Use sklearn to train a LogisticRegression model on the training set
logModel = LogisticRegression(max_iter=200)
logModel.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample = X.iloc[[0]]  
sampleProbs = logModel.predict_proba(sample)
sampleOutput = logModel.predict(sample)
print(f"Probabilities for each class: {sampleProbs}")
print(f"Predicted output: {sampleOutput}")

# iii. Report on the score for Logistic regression model, what does the score measure?
logScore = logModel.score(X_test, y_test)
print(f"Score: {logScore}")

# iv. Extract the coefficents and intercepts for the boundary line(s)
coefficients = logModel.coef_
intercept = logModel.intercept_
print(f"\nCoefficients: {coefficients}")
print(f"\nIntercept: {intercept}")

logScores = cross_val_score(logModel, X, y, cv=50)
print("\nLogistic Regression CV scores:", logScores)

Probabilities for each class: [[9.78174794e-01 2.18251395e-02 6.68170676e-08]]
Predicted output: ['setosa']
Score: 1.0

Coefficients: [[-0.37819275  0.91438454 -2.358225   -1.01764915]
 [ 0.52439024 -0.31503781 -0.20882288 -0.75351885]
 [-0.1461975  -0.59934673  2.56704788  1.771168  ]]

Intercept: [  9.0007707    1.83596076 -10.83673146]

Logistic Regression CV scores: [1.         1.         1.         1.         1.         1.
 0.66666667 1.         1.         1.         1.         1.
 1.         1.         1.         1.         1.         1.
 1.         0.66666667 0.66666667 1.         1.         1.
 1.         1.         1.         0.66666667 1.         1.
 1.         1.         1.         0.66666667 1.         1.
 1.         1.         1.         1.         1.         1.
 1.         1.         1.         1.         1.         1.
 1.         1.        ]


### REPORT

The score of the regression model is 1.0, which is a measure of the proportion of correctly classified samples out of the total samples in the test set. Since the score is 1.0, it means that it classified every single datapoint correctly. This high accuracy can suggest overfitting, especially since the size of the dataset is relatively small. 

The cross-validation scores show how for most cases, the model achieved perfect accuracy but for some folds, the score dropped to 0.667, suggesting that there is variability in the model's performance. 

Boundary line equations:

Class 0 (Setosa):

$$9.001 - 0.378 x_1 + 0.914 x_2 - 2.358 x_3 - 1.018 x_4 = 0$$

Class 1 (Versicolor):

$$1.836 - 0.524 x_1 - 0.315 x_2 - 0.209 x_3 - 0.754 x_4 = 0$$

Class 2 (Virginica):

$$-10.837 - 0.146 x_1 + 0.599 x_2 + 2.567 x_3 + 1.771 x_4 = 0$$




# Part 4: Support Vector Machine

In [81]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svmModel = svm.SVC(probability=True)
svmModel.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sampleProbs = svmModel.predict_proba(sample)
sampleOutput = svmModel.predict(sample)
print(f"Probabilities for each class: {sampleProbs}")
print(f"Predicted output: {sampleOutput}")

# iii. Report on the score for the SVM, what does the score measure?
svmScore = svmModel.score(X_test, y_test)
print(f"\nScore: {svmScore}")

svm_scores = cross_val_score(svmModel, X, y, cv=50)
print(f"\nCross validation scores: \n {svm_scores}")

Probabilities for each class: [[0.97281167 0.01747529 0.00971304]]
Predicted output: ['setosa']

Score: 0.9666666666666667

Cross validation scores: 
 [1.         1.         1.         1.         1.         1.
 0.66666667 1.         1.         1.         1.         1.
 1.         1.         1.         1.         1.         1.
 1.         1.         1.         1.         1.         1.
 1.         1.         0.66666667 0.66666667 1.         1.
 1.         1.         1.         0.66666667 1.         1.
 1.         1.         0.66666667 1.         1.         1.
 1.         1.         1.         1.         1.         1.
 1.         1.        ]


The score for the SVM model is 0.9667, which is a measure of the proportion of correctly classified samples in the test set. Similar to the logistic regression model, the SVM model's accuracy is also very high (very close to 1.0), indicated that it classified most datapoints corretcly, but could also be an indication overfitting. 

The cross-validation scores are also similar to the logistic regression model's case: for most folds, the model achieved a perfect accuracy of 1.0, but for some folds, the score dropped to 0.667, once again suggesting that there is variability in the model's performance. 


# Part 5: Neural Network

In [87]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
mlpModel = MLPClassifier(hidden_layer_sizes=1850, random_state=25, max_iter=1000)
mlpModel.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sampleProbs = mlpModel.predict_proba(sample)
sampleOutput = mlpModel.predict(sample)
print(f"Probabilities for each class: {sampleProbs}")
print(f"Predicted output: {sampleOutput}")

# iii. Report on the score for the Neural Network, what does the score measure?
mlpScore = mlpModel.score(X_test, y_test)
print(f"\nScore: {mlpScore}")

# iv: Experiment with different options for the neural network, report on your best configuration
mlp_scores = cross_val_score(mlpModel, X, y, cv=50)
print(f"\nCross validation scores: \n {mlp_scores}")

Probabilities for each class: [[9.99933201e-01 6.67993488e-05 9.03373646e-16]]
Predicted output: ['setosa']

Score: 1.0

Cross validation scores: 
 [1.         1.         1.         1.         1.         1.
 1.         1.         1.         1.         1.         1.
 1.         1.         1.         1.         1.         1.
 1.         1.         0.66666667 1.         0.66666667 1.
 1.         1.         1.         1.         1.         1.
 1.         1.         1.         0.33333333 1.         1.
 1.         1.         1.         1.         1.         1.
 1.         1.         1.         1.         1.         1.
 1.         1.        ]


Very similar to the SVM model, the score for the MLP Classifier is 0.9667, which is a measure of the proportion of correctly classified samples in the test set, indicating that it classified most datapoints corretcly, but could also be an indication overfitting. Also simialr to the SVM model, the cross-validation scores show that for most folds, the model achieved a perfect accuracy of 1.0, but for some folds, the score dropped to 0.667, once again suggesting that there is variability in the model's performance. 


When the hidden layer sizes is over ~1850, the model achievs a score of 1.0. And from when the size is 3 to around 1850, the score is consistently 0.9667. Changing the random_state parameter affected the scores in a very variable way since the variable affects the randomness of the model's weights initialization and the training and testing sets. For example (all for hidden_layer_sizes=3):

| random_state    | score |
| -------- | ------- |
| 5  | 0.967    |
| 10 | 0.933     |
| 15    | 0.967    |
|25 | 0.733|
|35 | 0.667|
|50 | 0.567|


One instance where the model resulted in 1.0 score, which is the "best" configuration (potential overfitting) is hidden_layer_sizes=1850, random_state=25, max_iter=1000

# Part 6: K-Nearest Neighbors

In [83]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
knnModel = KNeighborsClassifier(n_neighbors=5)
knnModel.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
knnOut = knnModel.predict(sample)
print(knnOut)

# iii. Report on the score for kNN, what does the score measure?
knnScore = knnModel.score(X_test, y_test)
print(knnScore)

knn_scores = cross_val_score(knnModel, X, y, cv=50)
print(knn_scores)

['setosa']
0.9666666666666667
[1.         1.         1.         1.         1.         1.
 0.66666667 1.         1.         1.         1.         1.
 1.         1.         1.         1.         1.         1.
 1.         0.66666667 0.66666667 1.         0.66666667 1.
 1.         1.         1.         1.         1.         1.
 1.         1.         1.         0.66666667 1.         1.
 1.         1.         1.         1.         1.         1.
 1.         1.         1.         1.         1.         1.
 1.         1.        ]


Very similar to the SVM model, the score for the KNN model is 0.9667, which is a measure of the proportion of correctly classified samples in the test set, indicating that it classified most datapoints corretcly, but could also be an indication overfitting. 

The cross-validation scores are also similar to the SVM model: for most folds, the model achieved a perfect accuracy of 1.0, but for some folds, the score dropped to 0.667, once again suggesting that there is variability in the model's performance. 


# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

All the models performed similarly - the logistic regression model had a perfect score of 1.0, and the rest had a very close score of 0.9667. This surprised me because I assumed there would be at least some variability between the models' performances, but I think they yielded perfect scores because the dataset is so small with only 150 datapoints, so it is very easy to learn the data and overfit. 